### DISCUSSION TITLE
Testing Direction

### DISCUSSION THREAD

**Discussion by chrisfw** [tensorbase/tensorbase#67] [General]
Hi Jin,

I find this project very exciting as it is bringing together or will bring together a variety of components/projects that were already of great interest to me, e.g. Apache Arrow, DataFusion, Ballista.  

I come from a Java background and I am not familiar with Rust _**yet**_, but I can certainly assist by testing out functionality and reporting any issues I find.

Do you have any particular area you would like users/testers of the tensorbase project to focus on in order to exercise the functionality at this point?  Also, do you have a more granular roadmap you can provide to highlight upcoming features that you plan to implement?

Regards,
Chris

**Comment by jinmingjian** [comment]
@chrisfw  Chris, thanks for your interests and starting a discussion!

The Java background is not the problem. There are many Rustaceans (Rust developers) from Java ecosystem. From application layer coding, the experience of Rust is almostly same to the Java. Of course, it is still need low-level thinkings to make the best use of one system language. However, if you are interested to these, you can always do best! 

We have setup internal harnesses to exercise TB. (When they become matured, we will release them as well.) If you are interested, I can pull you into the project. This is a great start point (A.K.A. your "Testing Direction") to help TB to grow to an industry strength product. 

For more your roadmap questions, this, in fact, indicates TB ideas has not well shown out which are scattered around the issues. I will try to enable the project panel to see if we can have a more clear big picture.

**Reply by chrisfw** [reply]
@jinmingjian Jin, thanks for your reply.

I have a copy of the latest Rust Programming Language book and I plan to spend some time getting familiar with the language.

Regarding the harnesses, I am interested but I fear I will need some time to get more familiar and comfortable with rust and the project source code, etc.

Related to the Road Map, I believe I saw mention made of perhaps exposing TB functionality via a mysql server protocol compatible endpoint.  There are a number of projects utilizing the same sort of approach using either mysql or postgres wire protocol to greatly expand the potential use of the project/product to standard BI tools, database tools, etc.
[Materialize](https://materialize.com/) (postgres), [Dolt](https://docs.dolthub.com/getting-started/sql)  (mysql), and [QuestDB](https://questdb.io/docs/reference/api/postgres/) (postgres) are 3 examples off the top of my head.  As I am sure you are aware, Dremio is another great example where an ODBC and JDBC driver provide great accessibility to the data repository from any application/tool that supports those two standards.  I am interested to know where this topic falls in terms of the TB roadmap. 

Thanks for any thoughts, insight or further information you can provide.

**Reply by jinmingjian** [reply]
@chrisfw  yes. They are in the roadmap. To expand the client ecosystem is indeed important for next step. MySql or PgSql wire protocol will be implemented ASAP.  

For harnesses, they are just for kinds of tests tools for TB. So the internal working of TB is intentionally not needed. For your want to provide testing help, it is great. You just run some tests from the queries against the TB server instance and report the bugs/unsupported queries to issues. Of course, it is also good to have a look for Rust and projects before more works. It depends on you. And I am glad to provide any help when you want.

**Reply by jinmingjian** [reply]
@chrisfw I recall there is a ClickHouse Native JDBC driver, which should be supported by TB now[1].  It would be great If you'd like do some tests and feedback the problems your found. If no problem, I hope we could add the support of Java JDBC driver in doc.

[1] https://github.com/housepower/ClickHouse-Native-JDBC

**Reply by chrisfw** [reply]
@jinmingjian , Yes, sure thing.  I can work on testing the JDBC driver and report back on the results.

**Reply by chrisfw** [reply]
@jinmingjian, I tested the JDBC driver and reported my results in a new issue, which I closed because I realized the Native driver you linked is not the official driver.  I will retest with the one you linked to.  

I also noticed that there is also a Clickhouse ODBC driver, so please let me know if/when you would like me to begin testing it.

**Reply by jinmingjian** [reply]
@chrisfw thanks for testing. The CH's official odbc[1] is http protocol based, which has not been supported by TensorBase now.

[1]  https://github.com/ClickHouse/clickhouse-odbc
